File: service/format.go
package service

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/dtonair/money-bot/domain"
)

func successText(tx domain.Transaction, usedAI bool) string {
	kind := "chi tiêu"
	if tx.Type == domain.TransactionIncome {
		kind = "thu nhập"
	}
	aiNote := ""
	if usedAI {
		aiNote = " (AI đã hỗ trợ hiểu tin nhắn)"
	}
	return fmt.Sprintf("✅ Đã lưu %s: %s - %s ₫.%s", kind, boundText(tx.Content(), 300), formatDong(tx.Amount), aiNote)
}

func duplicateText(tx domain.Transaction) string {
	return fmt.Sprintf("ℹ️ Giao dịch này đã được ghi trước đó: %s - %s ₫.", boundText(tx.Content(), 300), formatDong(tx.Amount))
}

func usageText() string {
	return strings.Join([]string{
		"🤷 Mình chưa hiểu giao dịch này.",
		"Vui lòng nhập dạng: ăn tối 150k pizza",
		"Thu nhập: thu lương 20tr tháng 7",
		"Báo cáo: /summary hoặc 'chi tiêu tháng này'",
	}, "\n")
}

func summaryUsageText() string {
	return strings.Join([]string{
		"🤷 Mình chưa hiểu tháng cần báo cáo.",
		"Ví dụ: /summary, /summary tháng 5, /summary 05/2026, /summary tháng trước.",
	}, "\n")
}

func formatSummary(summary domain.MonthlySummary) string {
	monthName := vietnameseMonthName(summary.Month)
	lines := []string{fmt.Sprintf("📊 Báo cáo %s %d:", monthName, summary.Year), ""}
	if summary.EntryCount == 0 {
		lines = append(lines, "📭 Chưa có dữ liệu cho tháng này.")
		return strings.Join(lines, "\n")
	}
	lines = append(lines,
		"💸 Tổng chi tiêu: "+formatDong(summary.TotalExpenses)+" ₫",
		"💰 Tổng thu nhập: "+formatDong(summary.TotalIncome)+" ₫",
		"⚖️ Cân bằng: "+formatDong(summary.Balance)+" ₫",
		"📝 Số giao dịch: "+strconv.Itoa(summary.EntryCount),
	)
	return strings.Join(lines, "\n")
}

func vietnameseMonthName(month time.Month) string {
	months := []string{"", "tháng một", "tháng hai", "tháng ba", "tháng tư", "tháng năm", "tháng sáu", "tháng bảy", "tháng tám", "tháng chín", "tháng mười", "tháng mười một", "tháng mười hai"}
	if month < time.January || month > time.December {
		return "tháng ?"
	}
	return months[int(month)]
}

func formatDong(amount int64) string {
	negative := amount < 0
	if negative {
		amount = -amount
	}
	s := strconv.FormatInt(amount, 10)
	parts := make([]string, 0, len(s)/3+1)
	for len(s) > 3 {
		parts = append([]string{s[len(s)-3:]}, parts...)
		s = s[:len(s)-3]
	}
	parts = append([]string{s}, parts...)
	out := strings.Join(parts, ".")
	if negative {
		return "-" + out
	}
	return out
}

func boundText(text string, max int) string {
	text = strings.TrimSpace(strings.Join(strings.Fields(text), " "))
	if max <= 0 {
		return ""
	}
	runes := []rune(text)
	if len(runes) <= max {
		return text
	}
	return string(runes[:max-1]) + "…"
}

File: src/service/format.ts
import { financialRatios, type FinancialSummary } from '../domain/financial_summary.ts';
import type { MonthlySummary } from '../domain/summary.ts';
import {
  type Transaction,
  TRANSACTION_INCOME,
  TRANSACTION_INVEST,
  TRANSACTION_SAVING,
  transactionContent,
} from '../domain/transaction.ts';

export function successText(transaction: Transaction, usedAI: boolean): string {
  const kind = transactionKind(transaction.type);
  const aiNote = usedAI ? ' (AI đã hỗ trợ hiểu tin nhắn)' : '';
  return `✅ Đã lưu ${kind}: ${boundText(transactionContent(transaction), 300)} - ${
    formatDong(transaction.amount)
  } ₫.${aiNote}`;
}

export function duplicateText(transaction: Transaction): string {
  return `ℹ️ Giao dịch này đã được ghi trước đó: ${
    boundText(transactionContent(transaction), 300)
  } - ${formatDong(transaction.amount)} ₫.`;
}

export function imagePreviewText(transaction: Transaction): string {
  return imagePreviewTextBatch([transaction]);
}

export function imagePreviewTextBatch(transactions: Transaction[]): string {
  if (transactions.length === 1) {
    const transaction = transactions[0]!;
    const kind = transactionKind(transaction.type);
    return `🖼️ Mình đọc được ${kind}: ${boundText(transactionContent(transaction), 300)} - ${
      formatDong(transaction.amount)
    } ₫.\nVui lòng kiểm tra trước khi lưu.`;
  }
  const lines = [`🖼️ Tìm thấy ${transactions.length} giao dịch:`, ''];
  let income = 0;
  let expense = 0;
  let invest = 0;
  let saving = 0;
  transactions.forEach((transaction, index) => {
    const kind = transactionKind(transaction.type, true);
    if (transaction.type === TRANSACTION_INCOME) income += transaction.amount;
    else if (transaction.type === TRANSACTION_INVEST) invest += transaction.amount;
    else if (transaction.type === TRANSACTION_SAVING) saving += transaction.amount;
    else expense += transaction.amount;
    lines.push(
      `${index + 1}. ${displayDate(transaction.date)} · ${kind} · ${
        boundText(transactionContent(transaction), 220)
      } — ${formatDong(transaction.amount)} ₫`,
    );
  });
  lines.push(
    '',
    `Tổng thu nhập: ${formatDong(income)} ₫`,
    `Tổng chi tiêu: ${formatDong(expense)} ₫`,
    `Tổng đầu tư: ${formatDong(invest)} ₫`,
    `Tổng tiết kiệm: ${formatDong(saving)} ₫`,
    '⚠️ Bấm xác nhận để lưu tất cả giao dịch trong danh sách.',
  );
  return lines.join('\n');
}

export function successBatchText(transactions: Transaction[]): string {
  return transactions.length === 1
    ? successText(transactions[0]!, false)
    : `✅ Đã lưu ${transactions.length} giao dịch vào Google Sheet.`;
}

export function duplicateBatchText(transactions: Transaction[]): string {
  return transactions.length === 1
    ? duplicateText(transactions[0]!)
    : `ℹ️ ${transactions.length} giao dịch này đã được ghi trước đó.`;
}

function displayDate(date: string | undefined): string {
  if (!date) return '??/??/????';
  const [year, month, day] = date.split('-');
  return `${day}/${month}/${year}`;
}

export function imageConfirmationUnavailableText(): string {
  return '⌛ Xác nhận ảnh không còn hiệu lực. Vui lòng gửi lại ảnh.';
}

export function usageText(): string {
  return [
    '🤷 Mình chưa hiểu giao dịch này.',
    'Vui lòng nhập dạng: ăn tối 150k pizza',
    'Thu nhập: thu lương 20tr tháng 7',
    "Báo cáo: /report hoặc 'chi tiêu tháng này'",
  ].join('\n');
}

export function aiUnrecognizedText(): string {
  return [
    '🤔 Mình không nhận ra một giao dịch rõ ràng trong tin nhắn (không thấy số tiền hoặc chiều thu/chi không rõ).',
    'Thử lại với số tiền cụ thể, ví dụ: ăn tối 150k hoặc thu lương 20tr.',
  ].join('\n');
}

export function aiUnavailableText(): string {
  return '😵 Mình tạm thời không xử lý được tin nhắn (AI không khả dụng). Vui lòng thử lại sau vài phút.';
}

export function aiInvalidResponseText(): string {
  return '🤖 Mình không đọc được phản hồi từ AI để ghi giao dịch. Vui lòng thử gửi lại tin nhắn.';
}

export function reportUsageText(): string {
  return [
    '🤷 Mình chưa hiểu tháng cần báo cáo.',
    'Ví dụ: /report, /report tháng 5, /report 05/2026, /report tháng trước.',
  ].join('\n');
}

/** @deprecated Use reportUsageText. */
export const summaryUsageText = reportUsageText;

export function formatFinancialSummary(summary: FinancialSummary): string {
  const lines = ['📊 Tổng quan tài chính — toàn thời gian', ''];
  if (summary.entryCount === 0) {
    lines.push('📭 Chưa có giao dịch để tổng hợp.');
    return lines.join('\\n');
  }
  lines.push(
    `💵 Tổng thu nhập: ${formatDong(summary.totalIncome)} ₫`,
    `💸 Tổng chi tiêu: ${formatDong(summary.totalExpenses)} ₫`,
    `🏦 Tổng tiết kiệm: ${formatDong(summary.totalSaving)} ₫`,
    `📈 Tổng đầu tư: ${formatDong(summary.totalInvest)} ₫`,
    `💰 Tiền mặt khả dụng*: ${formatDong(summary.cashAvailable)} ₫`,
    '',
    `📝 ${summary.entryCount} giao dịch`,
  );
  if (summary.firstTransactionDate && summary.lastTransactionDate) {
    lines.push(`📅 Dữ liệu: ${summary.firstTransactionDate} → ${summary.lastTransactionDate}`);
  }
  const ratios = financialRatios(summary);
  if (ratios.expenseToIncome !== undefined) {
    lines.push(
      '',
      `Chi tiêu / thu nhập: ${formatRatio(ratios.expenseToIncome)}`,
      `Tiết kiệm / thu nhập: ${formatRatio(ratios.savingToIncome!)}`,
      `Đầu tư / thu nhập: ${formatRatio(ratios.investToIncome!)}`,
    );
  }
  lines.push(
    '',
    '* Theo dữ liệu đã ghi: thu nhập - chi tiêu - đầu tư - tiết kiệm. Đây không phải số dư ngân hàng.',
  );
  return lines.join('\\n');
}

function formatRatio(value: number): string {
  return `${(value * 100).toFixed(1)}%`;
}

export function formatSummary(summary: MonthlySummary): string {
  const lines = [`📊 Báo cáo ${vietnameseMonthName(summary.month)} ${summary.year}:`, ''];
  if (summary.entryCount === 0) {
    lines.push('📭 Chưa có dữ liệu cho tháng này.');
    return lines.join('\n');
  }
  lines.push(
    `💸 Tổng chi tiêu: ${formatDong(summary.totalExpenses)} ₫`,
    `💰 Tổng thu nhập: ${formatDong(summary.totalIncome)} ₫`,
    `📈 Tổng đầu tư: ${formatDong(summary.totalInvest)} ₫`,
    `🏦 Tổng tiết kiệm: ${formatDong(summary.totalSaving)} ₫`,
    `⚖️ Còn lại: ${formatDong(summary.balance)} ₫`,
    `📝 Số giao dịch: ${summary.entryCount}`,
  );
  return lines.join('\n');
}

function transactionKind(type: Transaction['type'], titleCase = false): string {
  const kind = type === TRANSACTION_INCOME
    ? 'thu nhập'
    : type === TRANSACTION_INVEST
    ? 'đầu tư'
    : type === TRANSACTION_SAVING
    ? 'tiết kiệm'
    : 'chi tiêu';
  return titleCase ? kind[0]!.toLocaleUpperCase('vi-VN') + kind.slice(1) : kind;
}

export function vietnameseMonthName(month: number): string {
  const months = [
    '',
    'tháng một',
    'tháng hai',
    'tháng ba',
    'tháng tư',
    'tháng năm',
    'tháng sáu',
    'tháng bảy',
    'tháng tám',
    'tháng chín',
    'tháng mười',
    'tháng mười một',
    'tháng mười hai',
  ];
  return month >= 1 && month <= 12 ? months[month]! : 'tháng ?';
}

export function formatDong(amount: number): string {
  const negative = amount < 0;
  const digits = Math.abs(amount).toLocaleString('en-US', { useGrouping: false });
  const grouped = digits.replace(/\B(?=(\d{3})+(?!\d))/gu, '.');
  return negative ? `-${grouped}` : grouped;
}

export function boundText(text: string, max: number): string {
  const normalized = text.trim().split(/\s+/u).filter(Boolean).join(' ');
  if (max <= 0) return '';
  const characters = Array.from(normalized);
  return characters.length <= max ? normalized : `${characters.slice(0, max - 1).join('')}…`;
}
